cuda/btx_cudainterval_callbacks.cpp

```diff
@@ -0,0 +1,589 @@
+
+  // At device operation launch time, save the device the operation was launched on,
+  // keyed by the name, since this is not available in the device profiling events
+  std::unordered_map<hpt_function_name_t, CUdevice> hpt_function_name_to_dev;
+
+  // State for device event profiling
+  //
+  // The flow of events for event profiling is:
+  //
+  // 1. device operation entry (cuMemcpy, cuLaunch, etc)
+  // 2. event_profiling
+  // 3. device operation exit
+  // 4. unknown number of intervening events
+  // ...
+  // 5. event profiling results
+  // 
+  // The information needed to emit a thapi device event is spread out throughout
+  // these calls, so we need to maintain state between them.
+  // 
+  // Since 1 and 2 always happen one after another in a given thread, we just need
+  // to save the function name and launch ts keyed by {hostname, process, thread}.
+  // In (2) we can associate this information with the events, so when the results
+  // are finally available it can be looked up.
+  std::unordered_map<hpt_t, fn_ts_t> hpt_profiled_function_name_and_ts;
+  std::unordered_map<hp_event_t, fn_ts_t> hp_event_to_function_name_and_ts;
+
+  // Encapsulate complex context management state
+  CUDAContextManager context_manager;
+};
+
+using data_t = struct data_s;
+
+void set_entry_ts(data_t *data, hpt_t hpt, int64_t ts) {
+  data->entry_ts[hpt] = ts;
+}
+
+int64_t get_entry_ts(data_t *data, hpt_t hpt) {
+  return data->entry_ts.at(hpt);
+}
+
+std::string strip_event_class_name(const char *str) {
+  std::string temp(str);
+  std::smatch match;
+  std::regex_search(temp, match, std::regex(":(.*?)_?(?:entry|exit)?$"));
+
+  // The entire match is hold in the first item, sub_expressions 
+  // (parentheses delimited groups) are stored after.
+  assert( match.size() > 1 && "Event_class_name not matching regex.");
+
+  return match[1].str();
+}
+
+static void send_host_message(void *btx_handle, void *usr_data, int64_t ts,
+                              const char *event_class_name, const char *hostname, int64_t vpid,
+                              uint64_t vtid, bool err) {
+  std::string event_class_name_striped = strip_event_class_name(event_class_name);
+  const int64_t _start = get_entry_ts(static_cast<data_t *>(usr_data), {hostname, vpid, vtid});
+
+  btx_push_message_lttng_host(btx_handle, hostname, vpid, vtid, _start, BACKEND_CUDA,
+                              event_class_name_striped.c_str(), (ts - _start), err);
+}
+
+static void send_traffic_message(void *btx_handle, void *usr_data, int64_t ts,
+                                 const char *event_class_name, const char *hostname,
+                                 int64_t vpid, uint64_t vtid, bool err) {
+  std::string event_class_name_stripped = strip_event_class_name(event_class_name);
+  auto state = static_cast<data_t *>(usr_data);
+  hpt_t key = {hostname, vpid, vtid};
+  const int64_t _start = get_entry_ts(state, key);
```

Comment:
In `send _host` we don't use `state`, so please pick one. ( i prefer the smalest one :) )